internal/processor/processor.go

```diff
@@ -106,21 +117,38 @@ func (o *OperationProcessorImpl) processOperations() {
 	}
 }
 
-func (o *OperationProcessorImpl) processOperation(ctx context.Context, op types.Operation) {
-	if _, exist := o.runningOperations[op.Id]; exist {
-		xlog.Debug(ctx, "operation already running", zap.String("operation", op.String()))
+func (o *OperationProcessorImpl) processOperation(
+	ctx context.Context, op types.Operation,
+) {
+	if _, exist := o.runningOperations[op.GetId()]; exist {
+		xlog.Debug(
+			ctx, "operation already running",
+			zap.String("operation", types.OperationToString(op)),
+		)
 		return
 	}
-	o.runningOperations[op.Id] = true
+	o.runningOperations[op.GetId()] = true
 	o.wg.Add(1)
 	go func() {
 		defer o.wg.Done()
-		xlog.Debug(ctx, "start operation handler", zap.String("operation", op.String()))
+		xlog.Debug(
+			ctx, "start operation handler",
+			zap.String("operation", types.OperationToString(op)),
+		)
+		genericOp := types.GenericOperation{
+			Id:      op.GetId(),
+			Type:    op.GetType(),
+			State:   op.GetState(),
+			Message: op.GetMessage(),
+		}
 		result, err := o.handlers.Call(ctx, op)
 		if err != nil {
-			xlog.Error(ctx, "operation handler failed", zap.String("operation", op.String()))
+			xlog.Error(
+				ctx, "operation handler failed",
+				zap.String("operation", types.OperationToString(op)),
+			)
 		}
-		o.results <- OperationHandlerResult{old: op, new: result}
+		o.results <- OperationHandlerResult{old: &genericOp, new: result}
```

Comment:
It is more convenient to change given op inside handler, so here I propose to deep-copy old operation. It will be changed when we introduce void handlers